File: vrsctl/tests/terminal.py
"""Real CLI/PTY checks. Run after: cargo build -p vrsctl -p vrsd

Uses a temporary socket, an ephemeral node port, and no init scripts. Does not
connect to the user's runtime. Requires only Python's standard library (Unix).
"""
import errno
import fcntl
import os
from pathlib import Path
import pty
import re
import select
import shlex
import struct
import subprocess
import tempfile
import termios
import time
import unittest


ROOT = Path(__file__).resolve().parents[2]
BIN = ROOT / "target" / "debug"
VALUE = "((:name :echo :node \"alpha\" :interface ((ping x) (pong y))) (:name :clock :interface ((now))))"
EXPR = "'" + VALUE
PRETTY = ('((:name :echo\n  :node "alpha"\n'
          '  :interface ((ping x) (pong y)))\n'
          ' (:name :clock :interface ((now))))\n')


class TerminalTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # /tmp keeps the Unix socket below macOS's short path limit.
        cls.temp = tempfile.TemporaryDirectory(prefix="vrs-pretty-", dir="/tmp")
        cls.socket = str(Path(cls.temp.name) / "runtime.sock")
        cls.daemon = subprocess.Popen(
            [BIN / "vrsd", "--node", "format-test", "--node-port", "0", "--socket", cls.socket],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        deadline = time.monotonic() + 10
        while not Path(cls.socket).exists():
            if cls.daemon.poll() is not None or time.monotonic() > deadline:
                cls.tearDownClass()
                raise RuntimeError("Test daemon failed to start")
            time.sleep(0.02)

    @classmethod
    def tearDownClass(cls):
        cls.daemon.terminate()
        cls.daemon.wait(timeout=5)
        cls.temp.cleanup()

    def command(self, *args):
        return [str(BIN / "vrsctl"), "--socket", self.socket, *args]

    def pipe(self, *args, source=None):
        return subprocess.run(self.command(*args), input=source, text=True,
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                              timeout=5, check=True).stdout

    def terminal(self, *args, stdin=False, columns=40):
        master, slave = pty.openpty()
        fcntl.ioctl(slave, termios.TIOCSWINSZ, struct.pack("HHHH", 24, columns, 0, 0))
        proc = subprocess.Popen(self.command(*args),
                                stdin=slave if stdin else subprocess.DEVNULL,
                                stdout=slave, stderr=slave,
                                env={**os.environ, "TERM": "xterm-256color"})
        os.close(slave)

        def cleanup():
            if proc.poll() is None:
                # In the REPL, terminate before normal exit so no user history
                # file is saved. All runtime processes belong to our daemon.
                proc.terminate()
            proc.wait(timeout=5)
            os.close(master)
        self.addCleanup(cleanup)
        return proc, master

    def read_until(self, master, expected):
        output = b""
        deadline = time.monotonic() + 5
        while time.monotonic() < deadline:
            if select.select([master], [], [], 0.1)[0]:
                try:
                    chunk = os.read(master, 65536)
                except OSError as error:
                    if error.errno == errno.EIO:
                        break
                    raise
                if not chunk:
                    break
                output += chunk
                clean = re.sub(r"\x1b\[[0-?]*[ -/]*[@-~]", "", output.decode()).replace("\r", "")
                if expected in clean:
                    return clean
        self.fail(f"Expected {expected!r} in PTY output: {output!r}")

    def test_command_terminal_auto_width_and_compact_override(self):
        for args, expected in [((), PRETTY), (("--format", "compact"), VALUE + "\n")]:
            proc, master = self.terminal(*args, "-c", EXPR)
            self.read_until(master, expected)
            self.assertEqual(proc.wait(timeout=5), 0)

    def test_pipes_files_and_explicit_width(self):
        self.assertEqual(self.pipe("-c", EXPR), VALUE + "\n")
        self.assertEqual(self.pipe("--format", "pretty", "--width", "40", "-c", EXPR), PRETTY)
        self.assertEqual(self.pipe(source=EXPR), VALUE + "\n")
        source = Path(self.temp.name) / "values.ll"
        source.write_text(EXPR)
        self.assertEqual(self.pipe(str(source)), VALUE + "\n")
        proc, master = self.terminal(str(source))
        self.read_until(master, PRETTY)
        self.assertEqual(proc.wait(timeout=5), 0)
        proc, master = self.terminal("--width", "200", "-c", EXPR)
        self.read_until(master, VALUE + "\n")
        self.assertEqual(proc.wait(timeout=5), 0)

    def test_pretty_builtin_raw_strings_and_editor_comments(self):
        self.assertEqual(self.pipe("--raw", "-c", f"(pretty {EXPR} 40)"), PRETTY)
        self.assertEqual(self.pipe("--format", "pretty", "-c", r'"one\n\"two\""'), r'"one\n\"two\""' + "\n")
        transcript = self.pipe("--format", "editor", "--width", "45", "-c", EXPR)
        self.assertEqual(transcript.splitlines()[0], EXPR)
        self.assertTrue(all(line.startswith("# ") for line in transcript.splitlines()[1:]))

    def test_repl_pretty_and_compact(self):
        for args, expected in [((), PRETTY), (("--format", "compact"), VALUE + "\n")]:
            _, master = self.terminal(*args, stdin=True)
            self.read_until(master, "vrs> ")
            os.write(master, (EXPR + "\n").encode())
            self.read_until(master, expected)

    def test_quotation_round_trips_through_client_and_runtime(self):
        source = "(let ((name 'focus_window) (window '(:id 7))) `(continue_call ',name '(,window)))"
        self.assertEqual(self.pipe("-c", source), "(continue_call 'focus_window '((:id 7)))\n")
        self.assertEqual(self.pipe("-c", "'(unquote @name)"), ", @name\n")
        self.assertEqual(self.pipe("-c", f"(eq? (read (pretty {source} 8)) {source})"), "true\n")
        self.assertEqual(self.pipe("-c", "(let ((xs '(1 2))) `(a ,@xs))"), "(a 1 2)\n")

    def test_macros_expand_as_data_and_work_across_file_requests(self):
        self.assertEqual(self.pipe("-c", "(macroexpand_1 '(when! true (missing)))"),
                         "(if true (begin (missing)) nil)\n")
        source = "(defmacro plus_one (x) `(+ ,x 1))\n(plus_one! 41)\n"
        self.assertEqual(self.pipe(source=source), "plus_one\n42\n")
        self.assertEqual(self.pipe("-c", "(list (err? (try (plus_one! 1))) (when! true 42))"),
                         "(true 42)\n")

    def test_subscriptions_once_follow_and_clear(self):
        for index, mode in enumerate([(), ("-f",), ("-F",)]):
            topic = f"format_test_{index}"
            proc, master = self.terminal("-s", topic, *mode)
            # Publish repeatedly until the subscriber is ready; no arbitrary
            # assumption about process startup or subscription timing.
            deadline = time.monotonic() + 5
            while not select.select([master], [], [], 0.03)[0]:
                self.pipe("-c", f"(publish :{topic} {EXPR})")
                self.assertLess(time.monotonic(), deadline)
            self.read_until(master, PRETTY)
            if mode:
                self.pipe("-c", f"(publish :{topic} '(:second 42))")
                self.read_until(master, "(:second 42)\n")
                self.assertIsNone(proc.poll())
                fcntl.ioctl(master, termios.TIOCSWINSZ, struct.pack("HHHH", 24, 200, 0, 0))
                self.pipe("-c", f"(publish :{topic} {EXPR})")
                self.read_until(master, VALUE + "\n")
            else:
                self.assertEqual(proc.wait(timeout=5), 0)

    @unittest.skipUnless(os.environ.get("JANET_MODE_DIR"), "Set JANET_MODE_DIR to include Emacs integration")
    def test_emacs_evaluation_against_test_runtime(self):
        subprocess.run(
            ["emacs", "-Q", "--batch", "-L", os.environ["JANET_MODE_DIR"],
             "-L", str(ROOT / "emacs"), "-l", "lyric-mode-tests",
             "-f", "ert-run-tests-batch-and-exit"],
            env={**os.environ, "LYRIC_TEST_VRSCTL": shlex.join(self.command())},
            check=True, timeout=20)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: vrsctl/tests/terminal.rs
//! Real CLI/PTY checks against an isolated daemon.
//! Run `cargo build -p vrsd` before `cargo test -p vrsctl --test terminal`.
//! Use the same profile/target directory for both commands.

#![cfg(unix)]

mod support;

use std::fs;
use std::io;
use std::path::Path;
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{ensure, Result};
use assert_cmd::assert::OutputAssertExt;
use expectrl::Expect;
use support::{TestRuntime, TIMEOUT};

const VALUE: &str = "((:name :echo :node \"alpha\" :interface ((ping x) (pong y))) (:name :clock :interface ((now))))";
const PRETTY: &str = "((:name :echo\n  :node \"alpha\"\n  :interface ((ping x) (pong y)))\n (:name :clock :interface ((now))))\n";
const DEFAULT_PRETTY: &str = "((:name :echo :node \"alpha\" :interface ((ping x) (pong y)))\n (:name :clock :interface ((now))))\n";

#[test]
fn command_terminal_default_width_and_compact_override() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let expression = format!("'{VALUE}");
    for (args, expected) in [
        (vec!["-c", expression.as_str()], DEFAULT_PRETTY.to_owned()),
        (
            vec!["--format", "compact", "-c", &expression],
            format!("{VALUE}\n"),
        ),
    ] {
        let mut terminal = runtime.terminal(&args, false, 40)?;
        terminal.expect(&expected)?;
        terminal.success()?;
    }
    Ok(())
}

#[test]
fn pipes_files_and_explicit_width() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let expression = format!("'{VALUE}");
    assert_eq!(runtime.pipe(&["-c", &expression]), format!("{VALUE}\n"));
    assert_eq!(
        runtime.pipe(&["--format", "pretty", "--width", "40", "-c", &expression]),
        PRETTY
    );
    assert_eq!(
        runtime.pipe_input(&[], Some(&expression)),
        format!("{VALUE}\n")
    );
    let source = runtime.directory.path().join("values.ll");
    fs::write(&source, &expression)?;
    let source = source.to_str().unwrap();
    assert_eq!(runtime.pipe(&[source]), format!("{VALUE}\n"));
    let mut terminal = runtime.terminal(&[source], false, 40)?;
    terminal.expect(DEFAULT_PRETTY)?;
    terminal.success()?;
    let mut terminal = runtime.terminal(&["--width", "200", "-c", &expression], false, 40)?;
    terminal.expect(&format!("{VALUE}\n"))?;
    terminal.success()?;
    Ok(())
}

#[test]
fn pretty_builtin_raw_strings_and_editor_comments() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let expression = format!("'{VALUE}");
    assert_eq!(
        runtime.pipe(&["--raw", "-c", &format!("(pretty {expression} 40)")]),
        PRETTY
    );
    let string = r#""one\n\"two\"""#;
    assert_eq!(
        runtime.pipe(&["--format", "pretty", "-c", string]),
        format!("{string}\n")
    );
    let transcript = runtime.pipe(&["--format", "editor", "--width", "45", "-c", &expression]);
    let mut lines = transcript.lines();
    assert_eq!(lines.next(), Some(expression.as_str()));
    assert!(lines.all(|line| line.starts_with("# ")));
    Ok(())
}

#[test]
fn default_width_is_90_in_builtin_pipes_and_terminals() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let value = format!("({})", ["1234567890"; 8].join(" "));
    assert_eq!(value.len(), 89);
    assert_eq!(
        runtime.pipe(&["--raw", "-c", &format!("(pretty '{value})")]),
        format!("{value}\n")
    );
    assert_eq!(
        runtime.pipe(&["--format", "pretty", "-c", &format!("'{value}")]),
        format!("{value}\n")
    );
    for columns in [40, 200] {
        let mut terminal = runtime.terminal(&["-c", &format!("'{value}")], false, columns)?;
        terminal.expect(&format!("{value}\n"))?;
        terminal.success()?;
    }
    Ok(())
}

#[test]
fn repl_pretty_and_compact() -> Result<()> {
    let runtime = TestRuntime::new()?;
    for (args, expected) in [
        (vec![], DEFAULT_PRETTY.to_owned()),
        (vec!["--format", "compact"], format!("{VALUE}\n")),
    ] {
        let mut terminal = runtime.terminal(&args, true, 40)?;
        terminal.expect("vrs> ")?;
        terminal.session.send_line(format!("'{VALUE}"))?;
        terminal.expect(&expected)?;
        terminal.expect("vrs> ")?;
    }
    Ok(())
}

#[test]
fn quotation_round_trips_through_client_and_runtime() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let source =
        "(let ((name 'focus_window) (window '(:id 7))) `(continue_call ',name '(,window)))";
    assert_eq!(
        runtime.pipe(&["-c", source]),
        "(continue_call 'focus_window '((:id 7)))\n"
    );
    assert_eq!(runtime.pipe(&["-c", "'(unquote @name)"]), ", @name\n");
    assert_eq!(
        runtime.pipe(&["-c", &format!("(eq? (read (pretty {source} 8)) {source})")]),
        "true\n"
    );
    assert_eq!(
        runtime.pipe(&["-c", "(let ((xs '(1 2))) `(a ,@xs))"]),
        "(a 1 2)\n"
    );
    Ok(())
}

#[test]
fn macros_expand_as_data_and_work_across_file_requests() -> Result<()> {
    let runtime = TestRuntime::new()?;
    assert_eq!(
        runtime.pipe(&["-c", "(macroexpand_1 '(when! true (missing)))"]),
        "(if true (begin (missing)) nil)\n"
    );
    let source = "(defmacro plus_one (x) `(+ ,x 1))\n(plus_one! 41)\n";
    assert_eq!(runtime.pipe_input(&[], Some(source)), "plus_one\n42\n");
    assert_eq!(
        runtime.pipe(&["-c", "(list (err? (try (plus_one! 1))) (when! true 42))"]),
        "(true 42)\n"
    );
    Ok(())
}

#[test]
fn service_expansion_contains_direct_match_and_round_trips() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let source = "(begin
      (defn! echo (x) x)
      (def exports '(echo))
      (def code (macroexpand_1 '(srv! :test :interface exports)))
      (list (pretty code) (eq? code (read (pretty code)))
            (contains? (ls_srv) :test)))";
    let output = runtime.pipe(&["-c", source]);
    assert!(output.contains("((:echo x) (echo x))"), "{output}");
    assert!(
        output.contains(r#"(_ '(:err \"Unrecognized message\"))"#),
        "{output}"
    );
    assert!(!output.contains("service_dispatch"), "{output}");
    assert!(output.ends_with(" true false)\n"), "{output}");
    Ok(())
}

#[test]
fn init_service_and_nested_service_macros() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let source = "(begin
      (bind_srv :test_factory)
      (start_child :test_child 42)
      (bind_srv :test_child)
      (read_value))";
    assert_eq!(runtime.pipe(&["-c", source]), "42\n");
    Ok(())
}

#[test]
fn subscriptions_once_follow_and_clear() -> Result<()> {
    let runtime = TestRuntime::new()?;
    for (index, mode) in [None, Some("-f"), Some("-F")].into_iter().enumerate() {
        let topic = format!("format_test_{index}");
        let mut args = vec!["-s", topic.as_str()];
        args.extend(mode);
        let mut terminal = runtime.terminal(&args, false, 40)?;
        let deadline = Instant::now() + TIMEOUT;
        // Publish until output arrives instead of assuming the subscriber has
        // registered after an arbitrary startup delay. Matching retains output.
        while !terminal
            .session
            .is_matched(DEFAULT_PRETTY.replace('\n', "\r\n"))?
        {
            ensure!(Instant::now() < deadline, "subscriber did not become ready");
            runtime.pipe(&["-c", &format!("(publish :{topic} '{VALUE})")]);
            thread::sleep(Duration::from_millis(30));
        }
        terminal.expect(DEFAULT_PRETTY)?;
        if mode.is_some() {
            runtime.pipe(&["-c", &format!("(publish :{topic} '(:second 42))")]);
            terminal.expect("(:second 42)\n")?;
            terminal.assert_alive()?;
            terminal.resize(200)?;
            runtime.pipe(&["-c", &format!("(publish :{topic} '{VALUE})")]);
            terminal.expect(DEFAULT_PRETTY)?;
        } else {
            terminal.success()?;
        }
    }
    Ok(())
}

#[test]
fn emacs_evaluation_against_test_runtime() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let mut command = assert_cmd::Command::new("emacs");
    command
        .args(["-Q", "--batch", "-L"])
        .arg(Path::new(env!("CARGO_MANIFEST_DIR")).join("../emacs"))
        .args(["-l", "vrs-mode-tests", "-f", "ert-run-tests-batch-and-exit"])
        .env("VRS_TEST_VRSCTL", runtime.emacs_command())
        .timeout(Duration::from_secs(20));
    match command.output() {
        Err(error) if error.kind() == io::ErrorKind::NotFound => {
            eprintln!("skipping Emacs evaluation: Emacs is not installed");
        }
        result => {
            let output = result?;
            eprint!("{}", String::from_utf8_lossy(&output.stderr));
            output.assert().success();
        }
    }
    Ok(())
}

#[test]
fn dbg_views_share_real_recording_filters_and_editor_source_origins() -> Result<()> {
    let runtime = TestRuntime::new()?;
    let source = "(defn! twice (x) (+ x x))\n(dbg! (map '(2 3) twice))";
    let request =
        serde_json::json!({"source":source,"file":"/tmp/observe.ll","line":20,"column":1});
    runtime.pipe_input(&["--session"], Some(&format!("{request}\n")));
    let transcript = runtime.pipe(&["dbg", "--once", "--all", "--details"]);
    ensure!(transcript.contains("(map '(2 3) twice)"), "{transcript}");
    ensure!(transcript.contains("/tmp/observe.ll:21:7"), "{transcript}");
    ensure!(transcript.contains("# arg 1: 2"), "{transcript}");
    let json = runtime.pipe(&["dbg", "--once", "--json", "--at", "observe.ll:20:18"]);
    // Correct source location (the inner + starts at column 18 here).
    let history: serde_json::Value = serde_json::from_str(&json)?;
    let records = history["records"].as_array().unwrap();
    ensure!(
        records
            .iter()
            .filter(|r| r["site"]["form"] == "(+ x x)")
            .count()
            == 2,
        "{json}"
    );
    Ok(())
}
